envoy/config/xds_config_tracer.h

```diff
@@ -0,0 +1,111 @@
+#pragma once
+
+#include <string>
+#include <vector>
+
+#include "envoy/common/optref.h"
+#include "envoy/config/subscription.h"
+#include "envoy/config/typed_config.h"
+#include "envoy/protobuf/message_validator.h"
+
+#include "source/common/protobuf/protobuf.h"
+
+#include "google/rpc/status.pb.h"
+
+namespace Envoy {
+namespace Config {
+
+// The status of processing Delta/DiscoveryResponse.
+enum TraceState {
+  // Successfully got the resources or message.
+  RECEIVE = 0,
+  // Successfully ingested the resouces.
+  INGESTED = 1,
+  // Failed to apply the resources.
+  FAILED = 2,
+};
+
+struct TraceDetails {
+  TraceDetails(const TraceState state) : state_(state){};
+  TraceDetails(const TraceState state, const ::google::rpc::Status error_detail)
+      : state_(state), error_detail_(error_detail){};
+  const TraceState state_;
+  ::google::rpc::Status error_detail_;
+};
+
+/**
+ * An interface for hooking into xDS update events to provide the ablility to use some
+ * external logger or processor in xDS update. This tracer provides the log point when
+ * the resources are received , when they are successfully processed and applied, and when
+ * there is any failure.
+ *
+ * Instance of this interface get invoked on the main Envoy thread. Thus, it is important
+ * for implementations of this interface to not execute any blocking operations on the same
+ * thread.
+ */
+class XdsConfigTracer {
```

Comment:
As discussed in an external comment, this can probably be more generic any config processing operation (not only tracing/logging).

Reply:
Done.